scheduler/reconcile_util.go

```diff
@@ -258,6 +262,15 @@ func (a allocSet) filterByTainted(taintedNodes map[string]*structs.Node, serverS
 				if supportsDisconnectedClients {
 					// Filter running allocs on a node that is disconnected to be marked as unknown.
 					if alloc.ClientStatus == structs.AllocClientStatusRunning {
+						// Handle the case where the client updates its allocs
+						// before sending a heartbeat. The alloc will be set to
+						// running, but the client is still considered to be
+						// disconnected.
```

Comment:
I thought this was prevented by the node readiness check added in Node.UpdateAlloc?

Reply:
Yup, the RPC update is enough for this particular issue, but I was wondering if there was another scenario that we don't know about that could trigger this condition (like some data corruption? or bad alloc in state?) and so I kept the reconciler fix as well as a kind of layered defense. 

But I can remove this as well, less code to think about? 😅

Reply:
That's what I meant by the layered approach. The RPC change fixes things externally but without this change the reconciler can still _in theory_  generate an invalid result given a these specific conditions. 

I would've liked to be able to fix everything in the scheduler, but I found one case for system jobs that I couldn't and only the RPC ordering fixed it, and I suspect there may be other edge cases like this, so I kept both approaches.

But I agree that a comment stating this should not happen would be good.

Reply:
I'm sold now 😄 

Adding code to the scheduler that we don't expect to hit just add stuff to an already complex part of the code, so I removed this check and its equivalent for `system` scheduling. Thanks for the feedback!